source/server/connection_handler_impl.cc

```diff
@@ -77,16 +77,30 @@ ConnectionHandlerImpl::SslActiveListener::SslActiveListener(
                                                            parent.stats_store_, listener_options),
                      factory, socket.localAddress()->asString()) {}
 
-Network::Listener* ConnectionHandlerImpl::findListenerByPort(uint32_t port) {
-  // findListenerByPort is O(#listeners) operation, may need to add a map<port, listener>
-  // to improve performance to O(log #listeners); since the number of listeners is small
-  // linear performance might be adequate
-  auto l = std::find_if(
-      listeners_.begin(), listeners_.end(),
-      [port](const std::pair<Network::Address::InstancePtr, ActiveListenerPtr>& p) {
-        return p.first->type() == Network::Address::Type::Ip && p.first->ip()->port() == port;
-      });
-  return (l != listeners_.end()) ? l->second->listener_.get() : nullptr;
+Network::Listener*
+ConnectionHandlerImpl::findListenerByAddress(Network::Address::InstancePtr address) {
+  // This is a linear operation, may need to add a map<address, listener> to improve performance.
+  // However, linear performance might be adequate since the number of listeners is small.
+  auto listener =
+      std::find_if(listeners_.begin(), listeners_.end(),
+                   [address](const std::pair<Network::Address::InstancePtr, ActiveListenerPtr>& p) {
+                     return p.first->type() == Network::Address::Type::Ip && p.first == address;
+                   });
+
+  // If there is exact address match, return the corresponding listener.
+  if (listener != listeners_.end()) {
+    return listener->second->listener_.get();
+  }
+
+  // Otherwise, we need to look for the wild card match, i.e., 0.0.0.0:[address_port].
```

Comment:
It would be kind of nice to not have to do 2 searches here. I think this was commented on elsewhere, but wondering if we should do this in follow up. I wonder if we should have different options, "use_original_port" and "use_original_address". Not a huge deal, but throwing it out there.